src/vm/functions/database.rs

```diff
@@ -8,34 +8,97 @@ use vm::representations::{SymbolicExpression, SymbolicExpressionType};
+                        let contract_context_defining_trait = contract_defining_trait.contract_context;
+
+                        // Retrieve the function that will be invoked
+                        let function_to_check = contract_context_to_check.lookup_function(function_name)
+                            .ok_or(CheckErrors::BadTraitImplementation(trait_name.clone(), function_name.to_string()))?;
+                        
+                        // Check read/write compatibilty: if we are in a readonly context, then the next function should be readonly.
+                        // Otherwise the method should be public
+                        match (env.global_context.is_read_only(), &function_to_check.define_type) {
+                            (_, DefineType::Private) => return Err(CheckErrors::NoSuchPublicFunction(contract_identifier.to_string(), function_name.to_string()).into()),
+                            (true, DefineType::Public) => return Err(CheckErrors::WriteAttemptedInReadOnly.into()),
+                            (_, DefineType::ReadOnly) | (false, DefineType::Public) => { /* all clear */ },
+                        }
+
+                        function_to_check.check_trait_expectations(
+                            &contract_context_defining_trait,
+                            &trait_identifier,
+                            &contract_context_to_check)?;
+
+                        // Retrieve the expected method signature
+                        let constraining_trait = contract_context_defining_trait.lookup_trait_definition(&trait_name)
+                            .ok_or(CheckErrors::TraitReferenceUnknown(trait_name.clone()))?;
+                        let expected_sig = constraining_trait.get(function_name)
+                            .ok_or(CheckErrors::TraitMethodUnknown(trait_name.clone(), function_name.to_string()))?;
+                        (contract_identifier, Some(expected_sig.returns.clone()))
+                    }
+                },
+                _ => return Err(CheckErrors::ContractCallExpectName.into())
+            }
+        },
+        _ => return Err(CheckErrors::ContractCallExpectName.into())
+    };
 
     let contract_principal = Value::Principal(PrincipalData::Contract(
         env.contract_context.contract_identifier.clone()));
     let mut nested_env = env.nest_with_caller(contract_principal);
 
-    nested_env.execute_contract(&contract_identifier, 
-                                function_name, 
-                                &rest_args)
+    let result = nested_env.execute_contract(&contract_identifier, 
+                                             function_name, 
+                                             &rest_args)?;
+    
+    // Ensure that the expected type from the trait spec admits
+    // the type of the value returned by the dynamic dispatch.
+    if let Some(returns_type_signature) = type_returns_constraint {
```

Comment:
Shouldn't this be checked before calling the concretized function?

Reply:
Unfortunately we're not storing the contract analysis (unless I missed something), as such we don't know what is the return type of that function before the execution.

Reply:
Does the need for this check go away if we strictly require all concretizations of a trait to explicitly `impl-trait`?

Reply:
Yes, the check goes away, because the type check analysis already guaranteed that the inputs / output are consistent with the trait definition.

Reply:
I think we should flag this as part of the same todo as moving the dynamic check to the outermost call --- we can make this analysis information available to the runtime, it will just require some tweaking in the vm.